core/modules/replica/tools/qserv-replica-master.cc

```diff
@@ -26,191 +26,835 @@
+                                // Evict the worker
+
+                                LOGS(_log, LOG_LVL_INFO, healthMonitorContext << "DeleteWorkerJob");
+
+                                std::atomic<bool> finished(false);
+
+                                auto const job = DeleteWorkerJob::create(
+                                    worker,
+                                    permanentDelete,
+                                    controller,
+                                    parentJobId,
+                                    [&finished](DeleteWorkerJob::Ptr const& job) {
+                                        finished = true;
+                                    }
+                                );
+                                job->start();
+
+                                if (track(job, finished, "DeleteWorkerJob")) return;
+
+                                // -------------------------------------------------------------------------
+                                // Reset worker-non-response intervals before restart the Replication thread
+                                //
+                                // ATTENTION: the map needs to be rebuild from scratch because one worker
+                                // has been evicted from the Configuration.
+
+                                workerServiceNoResponseSec.clear();
+
+                                for (auto&& worker: controller->serviceProvider()->config()->workers()) {
+                                    workerServiceNoResponseSec[worker]["qserv"] = 0;
+                                    workerServiceNoResponseSec[worker]["replication"] = 0;
+                                }
+
+                                startReplicationSequence();
+                                break;
+                            }
+                            
+                            // Otherwise, proceed down to the default scenario.
+    
+                        default:
+
+                            // Any succesful replication effort is not possible at this stage due
+                            // to one of the following reasons (among other possibilities):
+                            //
+                            //   1) multiple nodes failed simultaneously
+                            //   2) all services on the worker nodes are down (typically after site outage)
+                            //   3) network problems
+                            //
+                            // So, we just keep monitoring the status of the system. The problem (unless it's
+                            // cases 2 or 3) should require a manual repair.
+
+                            LOGS(_log, LOG_LVL_INFO, healthMonitorContext
+                                 << "automated workers eviction is not possible because too many workers "
+                                 << workers2evict.size() << " are offline");
+
+                            break;
```

Comment:
Master Replication controller based on a fixed logic, and using the Cluster Health Monitor (a part of this project).